modules/loggers/mod_log_config.c

```diff
@@ -1666,13 +1781,279 @@ static ap_log_writer *ap_log_set_writer(ap_log_writer *handle)
+    apr_array_header_t *strs = apr_array_make(r->pool, lfd->nelts * 3, sizeof(char *)); /* array of pointers to char */
+    apr_array_header_t *strl = apr_array_make(r->pool, lfd->nelts * 3, sizeof(int));    /* array of int (strlen) */
+
+    lfdj->total_len = 0;
+
+    /* build json object */
+    lfdj->total_len += add_str(strs, strl, "{");
+    for (i = 0; i < lfd->nelts; ++i) {
+        if(items[i].tag == NULL) {
+            continue;
+        }
+
+        if (formatter_options->use_short_attribute_names) {
+            attribute_name = NULL;
+        }
+        else {
+            attribute_name = apr_hash_get(json_hash, items[i].tag, APR_HASH_KEY_STRING);
+        }
+
+        if (!attribute_name) {
+            attribute_name = ap_escape_logjson(r->pool, items[i].tag, NULL, 0); /* use tag as attribute name as fallback */
+        }
+
+        lfdj->total_len += add_str(strs, strl, "\"");
+        lfdj->total_len += add_str(strs, strl, attribute_name);
+        lfdj->total_len += add_str(strs, strl, "\":");
+
+        /* if the current tag has arguments, i.e. "i" (requestHeaders)
+         * create a sub object with items[i].arg as key and lfd->portions[i] values.
+         *
+         * as multiple same tags with different arguments can exist,
+         * the log_format_item is sorted before, so we can easily check for
+         * gruppenwechsel
+         */
+        if(items[i].arg != NULL && strlen(items[i].arg) > 0) {
+            /* start sub object */
+            lfdj->total_len += add_str(strs, strl, "{");
+            for (j = i; j < lfd->nelts; ++j) {
+                if(items[j].tag == NULL) {
+                    continue;
+                }
+                if(strcmp(items[j].tag, items[i].tag) != 0) {
+                    i = j; /* -1 ?*/
+                    break;
+                }
+                attribute_name = ap_escape_logjson(r->pool, items[i].arg, NULL, 0);
+
+                lfdj->total_len += add_str(strs, strl, "\"");
+                lfdj->total_len += add_str(strs, strl, attribute_name);
+                lfdj->total_len += add_str(strs, strl, "\":");
+
+                if (!lfd->portions[i]) {
+                    lfdj->total_len += add_str(strs, strl, "null");
+                }
+                else if (!*lfd->portions[i]) {
+                    lfdj->total_len += add_str(strs, strl, "\"\"");
+                }
+                else {
+                    lfdj->total_len += add_str(strs, strl, "\"");
+                    attribute_value = ap_escape_logjson(r->pool, lfd->portions[i], NULL, 0);
+                    lfdj->total_len += add_str(strs, strl, attribute_value);
+                    lfdj->total_len += add_str(strs, strl, "\"");
+                }
+                lfdj->total_len += add_str(strs, strl, ",");
+            }
+            /* remove last "," */
+            apr_array_pop(strs);
+            lfdj->total_len -= *(int *)apr_array_pop(strl);
```

Comment:
There is nothing to remove if all the items->tag were NULL in the `j` loop, is it possible?
Rather than removing maybe we could add if/when needed only, like:
```
            int found_j = 0;
            for (j = i; j < lfd->nelts; ++j) {
                if (items[j].tag == NULL) {
                    continue;
                }
                if (found_j) {
                    lfdj->total_len += add_str(strs, strl, ",");
                }
                else {
                    lfdj->total_len += add_str(strs, strl, "{");
                    found_j = 1;
                }
                ...
           }
```

Reply:
this should be fixed with latest commit which does implement omitting of null values.
also regarding your first question "There is nothing to remove if all the items->tag were NULL in the j loop, is it possible?" this should be also fixed with latest commit, but generall answer is: as all item[*].tags are sorted before usage in ap_json_log_formatter anyway, because of Gruppenwechsel logic, all NULL tags will be first and then later all non-nulls, so this check was in the code before the gruppenwechsel and sub object logic, and can probaly get removed now.